HpToolsLauncher/MtbxManager.cs

```diff
@@ -1,260 +1,275 @@
+ * constituting an additional warranty. Micro Focus shall not be liable for technical
+ * or editorial errors or omissions contained herein.
+ * The information contained herein is subject to change without notice.
+ * ___________________________________________________________________
+ *
+ */
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Xml.Linq;
+using System.Xml.Schema;
+using HpToolsLauncher.Properties;
+using HpToolsLauncher.TestRunners;
+
+namespace HpToolsLauncher
+{
+    public class MtbxManager
+    {
+
+        //the xml format of an mtbx file below:
+        /*
+         <Mtbx>
+            <Test Name="test1" path="${workspace}\test1">
+                <Parameter Name="mee" Value="12" Type="Integer"/>
+                <Parameter Name="mee1" Value="12.0" Type="Double"/>
+                <Parameter Name="mee2" Value="abc" Type="String"/>
+                <DataTable path="c:\tables\my_data_table.xls"/>
+                <Iterations mode="rngIterations|rngAll|oneIteration" start="2" end="3"/>
+            </Test>
+            <Test Name="test2" path="${workspace}\test2">
+                <Parameter Name="mee" Value="12" Type="Integer"/>
+                <Parameter Name="mee1" Value="12.0" Type="Double"/>
+                <Parameter Name="mee2" Value="abc" Type="String"/>
+            </Test>
+         </Mtbx>
+        */
+        public static List<TestInfo> LoadMtbx(string mtbxContent, string testGroup)
+        {
+            return LoadMtbx(mtbxContent, null, testGroup);
+        }
+
+        public static List<TestInfo> Parse(string mtbxFileName)
+        {
+            string xmlContent = File.ReadAllText(mtbxFileName);
+            return Parse(xmlContent, null, mtbxFileName);
+        }
+
+        private static XAttribute GetAttribute(XElement x, XName attributeName)
+        {
+            return x.Attributes().FirstOrDefault(a => a.Name.Namespace == attributeName.Namespace
+            && string.Equals(a.Name.LocalName, attributeName.LocalName, StringComparison.OrdinalIgnoreCase));
+        }
+
+        private static XElement GetElement(XElement x, XName eName)
+        {
+            return x.Elements().FirstOrDefault(a => a.Name.Namespace == eName.Namespace
+             && string.Equals(a.Name.LocalName, eName.LocalName, StringComparison.OrdinalIgnoreCase));
+        }
+
+        private static IEnumerable<XElement> GetElements(XElement x, XName eName)
+        {
+            return x.Elements().Where(a => a.Name.Namespace == eName.Namespace
+             && string.Equals(a.Name.LocalName, eName.LocalName, StringComparison.OrdinalIgnoreCase));
+        }
+
+        public static List<TestInfo> Parse(string mtbxFileName, Dictionary<string, string> jenkinsEnvironmentVars, string testGroupName)
+        {
+            string xmlContent = File.ReadAllText(mtbxFileName);

```

Comment:
As you checked the empty file content here, you can also add `try-catch` block to handle the **file read exception**, in case the user provided an invalid mtbx file.

Reply:
Agree, for various possible exceptions. When the file doesn't exist, code will not reach this point.